### PR TITLE
Add a test for TLS pipelining

### PR DESCRIPTION
**Pull request by mattcaswell**
TLS pipelining provides the ability for libssl to read or write multiple records in parallel. It requires special ciphers to do this, and there are currently no built-in ciphers that provide this capability. However, the dasync engine does have such a cipher, so we add a test for this capability using that engine.

##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by mattcaswell** [comment]
This went in surprisingly smoothly and easily. Given the length of time this code has been there without a test, and the recent refactor of the record layer - I was expecting problems.

**Comment by mattcaswell** [comment]
> The address sanitizer disagrees

Should be fixed now.

**Review by t8m — APPROVED** [review]
Approved with the typos fixed.

**Comment by mattcaswell** [comment]
Ping @openssl/committers for second review?

**Review by t-j-h — APPROVED** [review]

**Review by t8m — APPROVED** [review]

File: test/sslapitest.c
Comment (tmshort) [inline test/sslapitest.c:10173]: Cute, but perhaps something a little less cute, and a bit more obvious as to the length (yes, I counted it).

**Review by tmshort — COMMENTED** [review]
Mostly just a question on the use of ENGINE in the master branch, and whether this testing is sufficient (given the small data size).

File: test/sslapitest.c
Comment (mattcaswell) [reply test/sslapitest.c:10178]: > Are they fully deprecated

Yes

> Should this not be a provider or something else?

Yes, really this test *should* be using a provider. Unfortunately at the moment we are in the slightly strange situation that libssl pipeline support is not deprecated, but in order to use that support you must use a pipeline capable cipher - and we have not added the support for pipeline capable ciphers into providers. So the *only* way you can possibly use the non-deprecated libssl pipeline support is by using a deprecated ENGINE. I consider this an accidental omission which we should fix. But that goes way beyond what I was planning to do with this PR, i.e. just make sure what we've currently got isn't broken.

Comment (mattcaswell) [reply test/sslapitest.c:10178]: See also #18298

File: test/sslapitest.c
Comment (mattcaswell) [reply test/sslapitest.c]: Done

File: test/sslapitest.c
Comment (mattcaswell) [reply test/sslapitest.c]: I've added some more testing including one for large (maximal) pipelines.

**Comment by mattcaswell** [comment]
Adding the additional testing to this as suggested by @tmshort was a worthwhile exercise since it highlighted a couple of problems in the read pipelining implementation. I've fixed those issues and also made a tweak to the documentation as a result.

I suspect that the issues I fixed predated the record layer refactor so its possible those issues exist in 3.0 and 1.1.1 - but I've not investigated that.

**Comment by mattcaswell** [comment]
This now needs re-review since there have been some significant changes since the original version.

File: ssl/record/methods/tls_common.c
Comment (t8m) [inline ssl/record/methods/tls_common.c:293]: You can now drop the `{}`

File: doc/man3/SSL_CTX_set_split_send_fragment.pod
Comment (t8m) [inline doc/man3/SSL_CTX_set_split_send_fragment.pod:101]: Typo: significant

**Review by t8m — CHANGES_REQUESTED** [review]

File: test/sslapitest.c
Comment (tmshort) [reply test/sslapitest.c:10178]: "0-9A-Za-z" is 62 characters, perhaps?

**Comment by mattcaswell** [comment]
All feedback addressed. @t8m - please take another look.

**Review by t8m — APPROVED** [review]

**Comment by mattcaswell** [comment]
Ping for second review @openssl/committers

**Review by hlandau — APPROVED** [review]

File: ssl/record/methods/tls1_meth.c
Comment (tmshort) [inline ssl/record/methods/tls1_meth.c:165]: Why was this moved to the outer scope? It's the only change in this function.

File: ssl/record/methods/tls_common.c
Comment (tmshort) [inline ssl/record/methods/tls_common.c:307]: Does this special case no longer apply?

File: ssl/record/methods/tls1_meth.c
Comment (mattcaswell) [reply ssl/record/methods/tls1_meth.c:165]: This was due to use-after-scope error picked up by asan. The buffers set by `EVP_CTRL_SET_PIPELINE_OUTPUT_BUFS` must still be valid at the point that we call `EVP_CipherUpdate` - which was not the case when this was in its original location.

File: ssl/record/methods/tls_common.c
Comment (mattcaswell) [reply ssl/record/methods/tls_common.c:307]: This memmove was originally intended as an optimisation. It ensures that the payload data bytes are properly aligned for maximum assembler performance. There are two cases to consider: one where clearold is 0 and one where clearold is 1.

In the case that clearold is 0 this means "don't invalidate data earlier in the buffer because we're still using it". So this memove is just wrong in that instance. We must not move the data to an earlier position in the buffer because we're still using it. This actually was causing a problem in the pipelining case because we have multiple pipelines of data all using the same buffer - so this memmove was resulting in "bad record mac" error.

In the case that clearold is 1 it would still be a valid optimisation - but is entirely redundant  because of the memmove that occurs a few lines later. This *always* occurs if clearold is 1 and the data is not already aligned.

So, since this code is either wrong or redundant (depending on the value of clearold) it can safely be removed completely.

**Review by tmshort — APPROVED** [review]
Thanks for the explanations @mattcaswell

**Comment by t8m** [comment]
Merged to master branch. Thank you.

**Comment by mattcaswell** [comment]
> Merged to master branch. Thank you.

You beat me to it! I was just about to "git push"!

**Pull request closed without merging**
